### PR TITLE
Catalog & writer concurrency fixes

### PR DESCRIPTION
**Pull request by rocketbitz**
NewWriter now only requires a TimeBucketInfo to write to disk. This, along with the concurrency refactor of the catalog have resolved the write concurrency issue exposed by the combination of batch writes during backfill and the on-disk aggregation plugin.

### CODE REVIEW

File: catalog/catalog.go [hunk 1]
```diff
@@ -259,144 +263,132 @@ func (subDir *Directory) AddFile(newYear int16) (finfo_p *io.TimeBucketInfo, err
...
-func (d *Directory) GatherCategoriesFromCache() (catList map[string]int8) {
+func (d *Directory) GatherCategoriesFromCache() *sync.Map {
 	// Must be thread-safe for WRITE access
 	// Provides a map of categories contained within and below this directory. Will create the list cache if nil.
-	d.RLock()
-	catList = d.catList
-	d.RUnlock()
-	if catList == nil {
-		return d.gatherCategoriesUpdateCache()
+	needToUpdate := (d.catList == nil)
```
Comment (umitanuki) [inline catalog/catalog.go]: Are you aware of this change?  d.catList can be nil by the time of the return below and the caller will see "Symbol not in catalog"

**Comment by umitanuki** [comment]
Overall, I don't know if syncMap is improving the situation or making it worse.  The previous code was guarding the access to the map correctly.  The issue was outside of that.

File: catalog/catalog.go [hunk 2]
```diff
@@ -259,144 +263,132 @@ func (subDir *Directory) AddFile(newYear int16) (finfo_p *io.TimeBucketInfo, err
...
+			d.dataFile.Range(func(k, v interface{}) bool {
+				val, _ := list.Load(d.category)
+				m := val.(*sync.Map)
+				m.Store(strconv.Itoa(int(v.(*io.TimeBucketInfo).Year)), 0)
+				return true
+			})
 		}
 	}
-	catList := make(map[string]map[string]int, 0)
+	catList := &sync.Map{}
```
Comment (umitanuki) [inline catalog/catalog.go]: why does this have to be sync map?  I don't see how this is used from different goroutines

File: executor/writer.go [hunk 3]
```diff
@@ -4,84 +4,53 @@ import (
...
+	return &Writer{
+		root: rootCatDir,
+		tgc:  tgc,
+		tbi:  tbi,
+	}, nil
 }
+
 func (w *Writer) AddNewYearFile(year int16) (err error) {
-	w.FileInfoByYear[year], err = w.BaseDirectory.AddFile(year)
+	dir, err := w.root.GetOwningSubDirectory(w.tbi.Path)
```
Comment (umitanuki) [inline executor/writer.go]: Well, you need to block someone from changing the catalog directory tree concurrently.  That's the whole point

File: planner/planner.go [hunk 4]
```diff
@@ -236,7 +236,7 @@ func (q *query) Parse() (pr *ParseResult, err error) {
 	// Check to see that the categories in the query are present in the DB directory
 	CatList := q.DataDir.GatherCategoriesFromCache()
 	for key := range q.Restriction.GetRestrictionMap() {
-		if _, ok := CatList[key]; !ok {
+		if _, ok := CatList.Load(key); !ok {
```
Comment (umitanuki) [inline planner/planner.go]: a good example of syncMap not needed.  nobody else can touch this map anyway

**Comment by umitanuki** [comment]
I'd be a bit surprised if you don't see the issue anymore with this change.  I suggest reverting the change of syncmap.

File: catalog/catalog.go (hunk 1 above)
Comment (rocketbitz) [reply catalog/catalog.go]: that was a bad merge. will be fixed.

**Pull request merged**